[iOS] Fix Flyout layout

emaf opened the pull request:
### Description of Change

This was initially motivated by https://github.com/dotnet/maui/issues/17965, but expanding our tests I found many other issues.

Most of those were related to the usage of properties that were assumed to not include margin or safe area, but they actually did.

Changes:
- Both properties initially named `MeasuredHeaderViewHeightWithNoMargin` and `ArrangedHeaderViewHeightWithNoMargin` did include margin values, so those were renamed to `MeasuredHeaderViewHeightWithMargin` and `ArrangedHeaderViewHeightWithMargin`.
- `HeaderTopMargin` was wrongly assumed to be just top margin. If margin was not explicitly set, it represented the top safe area. This made us account for the safe area in cases were it shouldn't be accounted for.
- The ScrollView content inset was not considering `HeaderMinimumHeight`, so the content was ending up under the header in scenarios where the measured height was smaller than the minimum height.
- Layout header: all the changes are around not adding margin in either the height or Y offset of the HeaderView Frame to avoid accounting for it twice. For the Y offset it is already handled by MAUI's layout system based on the subviews.
- Layout content: similar to the header, we were accounting for maring twice. 
- Fixed SafeArea.Bottom value for the header, which should be 0 (instead of SafeArea.Left) since the header doesn't need to consider it.
- The MAUI `CollectionView` platform view was not found so all the calculations around `ScrollView` were ignored (i.e., content inset, collapsing header, etc.). Its platform view is not a `UIScrollView` directly, but instead `UICollectionViewControllerWrapperView -> Subviews -> UICollectionView`.
- For `CollectionView` we need to fix up the offset when scrolling, because we expect the offset to be declared in the same way as `ScrollView.ScrollY`, which is a negative offset based on the header height (relative to its Y position value).

Also expanded `FlyoutHeaderContentAndFooterAllMeasureCorrectly` Controls.DeviceTest to include scenarios with different margin settings and ScrollView. All these new combinations added over 100 test cases.

### Issues Fixed

Fixes https://github.com/dotnet/maui/issues/17965

<!--
Are you targeting main? All PRs should target the main branch unless otherwise noted.
-->

emaf committed:
[iOS] Fix Flyout Header breaking clicking over items

The Flyout Header height was being defined as `ArrangedHeaderViewHeightWithNoMargin` + `HeaderTopMargin`, but the `HeaderView.MeasuredHeight` (value on which `ArrangedHeaderViewHeightWithNoMargin` is based on) does include the margin so we were accounting for it twice and making the header go over the items below it, making them not clickable. When there is no margin, `HeaderTopMargin` represents the safe area and it should not be accounted for either.

Fixes https://github.com/dotnet/maui/issues/17965

emaf committed:
Rename confusing properties

Rename `ArrangedHeaderViewHeightWithNoMargin` and `MeasuredHeaderViewHeightWithNoMargin`, to `ArrangedHeaderViewHeightWithMargin` and `MeasuredHeaderViewHeightWithMargin`, since both include the top margin value: https://github.com/dotnet/maui/blob/main/src/Core/src/Layouts/LayoutExtensions.cs#L27.

jsuarezruiz reviewed (commented):
I think something is missing based on the build errors.

```
D:\a\_work\1\s\src\Controls\tests\DeviceTests\Elements\Shell\ShellFlyoutTests.iOS.cs(38,11): error CS0103: The name 'OpenFlyout' does not exist in the current context [D:\a\_work\1\s\src\Controls\tests\DeviceTests\Controls.DeviceTests.csproj::TargetFramework=net8.0-maccatalyst]
D:\a\_work\1\s\src\Controls\tests\DeviceTests\Elements\Shell\ShellFlyoutTests.iOS.cs(39,30): error CS0103: The name 'GetFlyoutPlatformView' does not exist in the current context [D:\a\_work\1\s\src\Controls\tests\DeviceTests\Controls.DeviceTests.csproj::TargetFramework=net8.0-maccatalyst]
D:\a\_work\1\s\src\Controls\tests\DeviceTests\Elements\Shell\ShellFlyoutTests.iOS.cs(45,58): error CS0103: The name 'GetSafeArea' does not exist in the current context [D:\a\_work\1\s\src\Controls\tests\DeviceTests\Controls.DeviceTests.csproj::TargetFramework=net8.0-maccatalyst]
    3 Error(s)
```

emaf committed:
Fix tests

emaf committed:
Fix safe area margin bottom value

emaf committed:
Fix margins

emaf committed:
Polish flyout header and content offsets

emaf committed:
Remove extra space

emaf committed:
Simplify content layout

emaf committed:
Small code cleanup

emaf committed:
Fix condition

emaf committed:
More tests

emaf committed:
Test fix

emaf committed:
Simplify test

emaf committed:
Add ArrangedHeaderViewHeightWithOutMargin

emaf committed:
Broken

emaf committed:
Fix scrollview inset

emaf committed:
Fix all tests

emaf committed:
Fix scroll view content inset and expand tests

emaf committed:
Fix new test cases for Android

emaf committed:
Skip FlyoutHeaderContentAndFooterAllMeasureCorrectly for Android since there is a bug that makes tests fail

emaf commented:
Skipping new test cases for Android since there is a bug that makes those fail. https://github.com/dotnet/maui/pull/19303 fixes that problem and re-enables the tests. It's a draft PR just because it depends on this one and needs to be rebased once this is merged so it only contains the Android specific fix.

PureWeen reviewed (changes requested):
It looks like when you set Flyout Header Behavior to scroll it's also collapsing while it scrolls

![image](https://github.com/dotnet/maui/assets/5375137/58899e54-da77-4b8b-abec-6551e45f79b7)

After scroll

![image](https://github.com/dotnet/maui/assets/5375137/a2609b9e-ffa7-48a7-b84b-edc9c5c21151)

emaf commented:
> It looks like when you set Flyout Header Behavior to scroll it's also collapsing while it scrolls
> 
> ![image](https://private-user-images.githubusercontent.com/5375137/290009308-58899e54-da77-4b8b-abec-6551e45f79b7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI2NTc3ODIsIm5iZiI6MTcwMjY1NzQ4MiwicGF0aCI6Ii81Mzc1MTM3LzI5MDAwOTMwOC01ODg5OWU1NC1kYTc3LTRiOGItYWJlYy02NTUxZTQ1Zjc5YjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMTVUMTYyNDQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWZmZmFkMTgzYTQ5MzNmMDhhODJjYmE0M2ZjOWRmNzY4YTkyOTdmM2QyOTJjNWRmZWEzODdlMmU0MzRmODY0YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.2wx4n8Xe8uEDo4beMBF6EhyBhHsYl_40IR72lnKBECI)
> 
> After scroll
> 
> ![image](https://private-user-images.githubusercontent.com/5375137/290009272-a2609b9e-ffa7-48a7-b84b-edc9c5c21151.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI2NTc3ODIsIm5iZiI6MTcwMjY1NzQ4MiwicGF0aCI6Ii81Mzc1MTM3LzI5MDAwOTI3Mi1hMjYwOWI5ZS1mZmE3LTQ4YTctYjg0Yi1lZGM5YzVjMjExNTEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMTVUMTYyNDQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTIwNDlhZGU4ODIzNWI3NzgwZGMxZjdhOTAzZmUyYzZjZTc1Y2E0NTQ0YjcwOTc5MzFmNzBmMDYxNDM2ZTQ3NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.dK8PMVTy-TOx5QKd1CqLpWnVeW7qVjdEtge6KVfzGlc)

This is now fixed by https://github.com/dotnet/maui/pull/19125/commits/52c5ae99769c7ead1a96fe5a15c4f6c8fa7eb1b6

emaf commented:
> This is now fixed by [52c5ae9](https://github.com/dotnet/maui/commit/52c5ae99769c7ead1a96fe5a15c4f6c8fa7eb1b6)

Removed that commit. After chatting with @PureWeen, this seems to be the right behavior for now, since we don't have a bindlable property that would let us know if the `IgnoreSafeArea` property is explicitely set.

emaf committed:
Enable more tests for iOS

emaf committed:
Honor IgnoreSafeArea

emaf committed:
Add missing null check

emaf committed:
Fix CollectionView support

emaf committed:
Add check by ItemsView for CV

emaf committed:
Add tests scrolling using different content control types

emaf committed:
Log more info on new tests

emaf committed:
Use requested value for assertion

emaf committed:
Consider epsilon for scrolled position

emaf committed:
Better test logging

emaf committed:
Fix spacing

emaf committed:
Another indenting issue

emaf committed:
Fix namespaces

PureWeen commented:
/rebase
